### PR TITLE
Nschoe/draft

### PR DESCRIPTION
**Pull request by GoNominal**
A few types and functions to kickstart the lib' .
Nothing is safe, impl is absolutely trivial so far

### CODE REVIEW

**Comment by GoNominal** [comment]
`miden-vm` bumped to version `0.5.0` to support `is_odd`.

`Expr` type refactored to internal, hidden `Expr_`. Now `Expr a`, which is typed, is exposed.
"smart constructors" are now "smart": you can't create a "wrong `Expr`".

While block now supported.

Private inputs, too. Though the implementation is fairly naive for now (basically a `NextSecret` getter to access the next positional secret argument). This might overfit `miden`.

**Comment by GoNominal** [comment]
Add utility functions `inc`, `dec`, `withVarF`, etc. (for convenience, while we don't have `do` notation yet).
Added 32-bit integer division `IDiv32`.

File: src/Nightfall/Lang/Internal/Types.hs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    | NakedCall FunName [Expr_]
+
+    -- | Return
+    | Return (Maybe Expr_)
+
+    -- | Comment
+    | Comment String
+
+    -- | Allow to add empty lines in the generated code, for clarity
+    | EmptyLine
```
Comment (wingedtrashcan) [inline src/Nightfall/Lang/Internal/Types.hs:68]: I think I'd put all meta information like comments and empty lines into one constructor, so that it doesn't get in the way when processing parts of the AST that have actual semantics. Just a nitpick.

Also, we can't have comments inside of expressions? And we can't lay out an expression across several lines with some empty lines in between etc?

File: src/Nightfall/Lang/Types.hs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+lit :: Word64 -> Expr Felt
+lit = Expr . Lit
+
+bool :: Bool -> Expr Bool
+bool = Expr . Bo
+
+-- ** Arithmetic operations
+
+add :: Num a => Expr a -> Expr a -> Expr a
+add  (Expr e1) (Expr e2) = Expr $ Add e1 e2
```
Comment (wingedtrashcan) [inline src/Nightfall/Lang/Types.hs:109]: Why do you need the `Num` constraint here and other constraints below?

File: src/Nightfall/Lang/Types.hs [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+declareVarB varname (Expr e) = Statement $ DeclVariable varname e
+
+assignVarF :: VarName -> Expr Felt -> Statement
+assignVarF varname (Expr e) = Statement $ AssignVar varname e
+
+assignVarB :: VarName -> Expr Bool -> Statement
+assignVarB varname (Expr e) = Statement $ AssignVar varname e
+
+ifElse :: Expr Bool -> [Statement] -> [Statement] -> Statement
+ifElse (Expr cond) ifBlock elseBlock = Statement $ IfElse cond (coerce ifBlock) (coerce elseBlock)
```
Comment (wingedtrashcan) [inline src/Nightfall/Lang/Types.hs:186]: Horrible nitpick: should it be `thenBlock` rather than `ifBlock`?

File: src/Nightfall/Targets/Miden.hs [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+    midenInstr <- concat <$> mapM (transpileStatement . coerce) (pStatements zkProg)
+    return $ Module { moduleImports = [] -- No import for now (TODO)
+                    , moduleProcs = Map.empty -- No procs either (TODO)
+                    , moduleProg = Program (head warning : midenInstr)
+                    }
+
+transpileStatement :: Statement_ -> State Context [Instruction]
+transpileStatement (NFTypes.Comment str) = return . singleton . MASM.Comment . Text.pack $ str
+transpileStatement (IfElse cond ifBlock elseBlock) = do
+    ifBlock' <- concat <$> mapM transpileStatement ifBlock
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:69]: Given that `concat <$> mapM f xs` appears a number of times in this module, I'd give it a name, especially given that it does have an established name ([`foldMapA`](https://hackage.haskell.org/package/relude-1.2.0.0/docs/Relude-Foldable-Fold.html#v:foldMapA)).

File: src/Nightfall/Targets/Miden.hs [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    return $ cond' <> [ MASM.While $ body' <> cond' ]
+-- | Declaring a variable loads the expression into Miden's global memory, and we track the index in memory
+transpileStatement (DeclVariable varname e) = do
+    vars <- gets variables
+    pos <- gets memPos
+    cfg <- gets config
+
+    -- Check if this variable hasn't already been declared
+    when (Map.member varname vars) $ do
+        error $ "Variable \"" ++ varname ++ "\" has already been declared!"
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:87]: We don't allow shadowing? Does this restriction come from Miden? If we're going to unroll loops or something, not allowing shadowing might complicate things.

File: src/Nightfall/Targets/Miden.hs [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+    
+    -- Insert the record
+    let vars' = Map.insert varname pos vars
+
+    -- Update the context
+    modify $ \s -> s { memPos = pos + 1, variables = vars' }
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:93]: Can a variable go out of scope?

**Review by wingedtrashcan — COMMENTED** [review]
An initial shallow review. I'll need to get comfortable with Miden in order to be able to review transpilation properly.

File: README.md [hunk 7]
```diff
@@ -32,15 +32,36 @@ Before you begin, ensure you have the following software installed on your syste
...
+For now, the [Main](./app/Main.hs) is simply a dummy program to test the [examples](./src/Examples/). Launch with
+```sh
+cabal run nightfall -- <example-name>
+```
+
+to get the MASM output on stdin, or
 
 ```sh
-nix-shell --run "cabal run"
+cabal run nightfall -- <example-name> <filepath>
```
[the remark this answers is not in the document]
Comment (GoNominal) [reply README.md:49]: I guess it's a matter of preferences. We're trying to keep it work both ways. I'm mainly running it with `cabal` myself.

File: app/Main.hs [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+                                , ("simpleInf", simpleInfProg)
+                                , ("sumTo10", sumTo10Prog)
+                                , ("collatzFixed", collatzFixedProg)
+                                , ("collatzPriv", collatzPrivProg)
+                                , ("simpleSecret", simpleSecretProg)
+                                ]
+    
+    args <- getArgs
+
+    -- For now, trivial: argument is the name of the example to run
```
[the remark this answers is not in the document]
Comment (GoNominal) [reply app/Main.hs:35]: Oh yes, definitely. 
I just threw this "quick'n'dirty" to try.

This was just a way to test the lib, the actual executable (later) should take a program as input and produce the transpilation ; not run some pre-defined, hardcoded examples like right now.

File: src/Nightfall/Lang/Internal/Types.hs [hunk 9]
```diff
@@ -0,0 +1,69 @@
+module Nightfall.Lang.Internal.Types where
+
+import Data.Word ( Word64 )
+
+-- | Field Element type
+type Felt = Word64
```
[the remark this answers is not in the document]
Comment (GoNominal) [reply src/Nightfall/Lang/Internal/Types.hs:6]: Yes absolutely, I thought I had put a comment there to warn. But that might be in another branch or something.
Basically, since miden is going to be our primary backend "for a while", I used `Word64` because miden's field elements are slightly less than 64 bits (I think `p = 2^64 - 2^32 + 1` in case of Miden), so "it fits".

But it won't work with `Cairo` for instance which has a prime in the `2^256` range IIRC.

This is definitely something we must fix to support more backends ; you're correct.

File: src/Examples/Simple.hs [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+-- | That's the Haskell program to describe with the EDSL
+{-
+trivial3 :: Felt
+trivial3 = 1238 * (345 + 78)
+-}
+
+-- | That's the EDSL version
+trivial3Stmts :: [Statement]
+trivial3Stmts = [ comment "Perform an addition followed by a multipication, 1238 * (345 + 78), should output 523674"
+                , ret . Just $ mul (lit 1238) (add (lit 345) (lit 78))
```
[the remark this answers is not in the document]
Comment (GoNominal) [reply src/Examples/Simple.hs:66]: We can (it's the next `trivial4Stmts`).
The idea was to build examples "from scratch" in `Simple.hs`. I wanted to make a small introductory article explaining how to use `nifghtfall`. So going from completely trivial examples to less and less trivial.

Basically, "example 3" was to make sure the orders the elements were pushed on the stack when you have both an addition (`add`) and a multiplication (`mul`) were correct.

Then "example 4" would show how to leverage the `Num` instance, precisely to "remove the clutter". And then "from now one" (all the next examples) were using the uncluttered version.


TL;DR : this was just made for introductory purposes, for when/if we write a sort of "tutorial" for nightfall.

File: src/Nightfall/Lang/Internal/Types.hs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    | IDiv32 Expr_ Expr_ -- ^ a `quot` b with a and b 32-bit integers
+
+    -- | Boolean operations
+    | Equal Expr_ Expr_     -- ^ a == b
+    | Not Expr_             -- ^ !a
+    | Lower Expr_ Expr_     -- ^ a < b
+    | LowerEq Expr_ Expr_   -- ^ a <= b
+    | Greater Expr_ Expr_   -- ^ a > b
+    | GreaterEq Expr_ Expr_ -- ^ a >= b
+    | IsOdd Expr_           -- a `mod` 2 == 1
```
[the remark this answers is not in the document]
Comment (GoNominal) [reply src/Nightfall/Lang/Internal/Types.hs:33]: I'm happy to discuss this: you probably have more experience in eDSL design that I do.

This gets "even worse" later because we need to implement those operations for `Int32`, `Int64`, etc.
There must probably be a better way of doing it.

File: src/Nightfall/Lang/Internal/Types.hs (hunk 1 above)
Comment (GoNominal) [reply src/Nightfall/Lang/Internal/Types.hs:68]: > I think I'd put all meta information like comments and empty lines into one constructor, so that it doesn't get in the way when processing parts of the AST that have actual semantics. Just a nitpick.

Sure, I have no issues with this. I was experimenting: I was generating "relatively large" miden program for testing, and it was becoming a bit hard to follow by hand. So I wanted to introduce comments and new lines to make easier to inspect final transpiled program.

Totally okay to refactor if you have a more compact and smarter way 👍🏻

File: README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (wingedtrashcan) [reply README.md:49]: Thanks! It's a pity I don't have the rights to be able to resolve conversations.

File: app/Main.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (wingedtrashcan) [reply app/Main.hs:35]: > I just threw this "quick'n'dirty" to try.

I'd do it this way too probably. No matter how many times I used `optparse-applicative`, I still have to relearn it completely from scratch every single time.

File: src/Examples/Simple.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (wingedtrashcan) [reply src/Examples/Simple.hs:66]: Fair enough, thanks!

File: src/Nightfall/Lang/Internal/Types.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (wingedtrashcan) [reply src/Nightfall/Lang/Internal/Types.hs:33]: > This gets "even worse" later because we need to implement those operations for Int32, Int64, etc.
> There must probably be a better way of doing it.

The most straightforward one would be to create one AST node for all 1-ary functions (and make those into an enum), one AST node for all 2-ary functions etc. That would reduce the amount of boilerplate dramatically. It's also possible to generalize over the arity of functions and reduce the overhead further, but that requires GADTs and whatnot and may not be worth it. You can also do

```haskell
    | BuiltinApp Builtin [Expr_]
```

and lose static guarantees on the arity of a builtin matching the number of arguments it has in the AST.

File: src/Nightfall/Lang/Internal/Types.hs (hunk 1 above)
Comment (wingedtrashcan) [reply src/Nightfall/Lang/Internal/Types.hs:68]: > I was experimenting: I was generating "relatively large" miden program for testing, and it was becoming a bit hard to follow by hand. So I wanted to introduce comments and new lines to make easier to inspect final transpiled program.

Yeah, I do agree having comments in there is the right design.

> Totally okay to refactor if you have a more compact and smarter way 👍🏻

As I said, I'm just nitpicking, it isn't really important.

> Indeed, for now we can't. We'd have to check if miden support it (or whatever backend)

Surprisingly, it seems like it may not. Weird.

File: src/Nightfall/Lang/Types.hs (hunk 2 above)
Comment (wingedtrashcan) [reply src/Nightfall/Lang/Types.hs:109]: > Does it make sense?

`add True 67` is ruled out by the type system and the constraint is not needed, but `add True False` isn't ruled out by the type system without the constraint, hence I agree that adding the constraint improves type safety. Thanks for explaining!

File: app/Main.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (allezlom13) [reply app/Main.hs:35]: Same...

File: src/Nightfall/Lang/Internal/Types.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (allezlom13) [reply src/Nightfall/Lang/Internal/Types.hs:33]: If you have a smart constructor for each specific builtin, you can "recover" precise types I guess?

File: src/Nightfall/MASM/Types.hs [hunk 12]
```diff
@@ -59,13 +59,21 @@ data Instruction
...
   | MemLoad (Maybe Word32) -- mem_load[.i]
   | MemStore (Maybe Word32) -- mem_store[.i]
   | Add (Maybe Word32) -- add[.n]
   | Sub (Maybe Word32) -- sub[.n]
+  | Mul (Maybe Word32) -- mul[.n]
+  | Div (Maybe Word32) -- div[.n]
+  | Neg
   | IAdd -- u32checked_add
   | ISub -- "u32checked_sub"
   | IMul -- u32checked_mul
```
Comment (wingedtrashcan) [inline src/Nightfall/MASM/Types.hs:79]: Why no `Maybe Int32`?

File: src/Nightfall/MASM/Types.hs [hunk 13]
```diff
@@ -59,13 +59,21 @@ data Instruction
...
   | Not -- not
+  | IsOdd -- is_odd
   | LocStore Word32 -- loc_store.i
   | LocLoad Word32 -- loc_load.i
   | MemLoad (Maybe Word32) -- mem_load[.i]
   | MemStore (Maybe Word32) -- mem_store[.i]
   | Add (Maybe Word32) -- add[.n]
   | Sub (Maybe Word32) -- sub[.n]
+  | Mul (Maybe Word32) -- mul[.n]
+  | Div (Maybe Word32) -- div[.n]
```
Comment (wingedtrashcan) [inline src/Nightfall/MASM/Types.hs:75]: Not sure if I'm being crazy, but I thought those were field operations? I.e. defined over `Felt`, which is `Word64` rather than `Word32`?

BTW, it would be nice for `Felt` to be a `newtype`, so that the distinction between zk-native and zk-encoded stuff is apparent.

**Review by wingedtrashcan — COMMENTED** [review]
Another small chunk.

File: src/Nightfall/Targets/Miden.hs [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+                    }
+
+transpileStatement :: Statement_ -> State Context [Instruction]
+transpileStatement (NFTypes.Comment str) = return . singleton . MASM.Comment . Text.pack $ str
+transpileStatement (IfElse cond ifBlock elseBlock) = do
+    ifBlock' <- concat <$> mapM transpileStatement ifBlock
+    elseBlock' <- concat <$> mapM transpileStatement elseBlock
+    -- Not entirely sure, Miden talks about a "a small, but non-negligible overhead", but they don't say.
+    -- I'm supposing it takes a pop/drop operation + a comparison
+    cond' <- transpileExpr cond
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:74]: Hm, I think whatever we do, we should provide direct access to Miden's primitives in addition to our nice abstract interface. What if the right value is already at the top of the stack and the user doesn't want to create any expressions implicitly pushing stuff onto the stack and wants to just use `if.true` directly. Not exposing such functionality is guaranteed to annoy people who want/need to write efficient code, which I bet are going to be a sizeable chunk of our users.

File: src/Nightfall/Targets/Miden.hs [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+                    , moduleProg = Program (head warning : midenInstr)
+                    }
+
+transpileStatement :: Statement_ -> State Context [Instruction]
+transpileStatement (NFTypes.Comment str) = return . singleton . MASM.Comment . Text.pack $ str
+transpileStatement (IfElse cond ifBlock elseBlock) = do
+    ifBlock' <- concat <$> mapM transpileStatement ifBlock
+    elseBlock' <- concat <$> mapM transpileStatement elseBlock
+    -- Not entirely sure, Miden talks about a "a small, but non-negligible overhead", but they don't say.
+    -- I'm supposing it takes a pop/drop operation + a comparison
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:73]: I'd make it `IfElse Expr_ [Statement_] (Maybe [Statement_])`, because imposing limitations on the underlying system without a specific reason for that is going to annoy users.

File: src/Nightfall/Targets/Miden.hs [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+        traceVar = [MASM.Comment $ "var " <> Text.pack varname | shouldTrace]
+    
+    e' <- transpileExpr e
+
+    return $ e' <> traceVar <> [ MASM.MemStore . Just . fromIntegral $ pos ]
+
+-- | A (naked) function call is done by pushing the argument on the stack and caling the procedure name
+transpileStatement (NakedCall fname args) = do
+    args' <- concat <$> mapM transpileExpr args
+    return $ args' <> [ MASM.Exec . Text.pack $ fname ]
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:130]: We should probably use `Text` in constructors consistently.

File: src/Nightfall/Targets/Miden.hs [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+    -- When we do have a value, we simply push it to the stack
+    Just e -> transpileExpr e
+
+transpileStatement EmptyLine = return . singleton $ MASM.EmptyL
+
+transpileStatement _ = error "transpileStatement::TODO"
+
+
+-- TODO: range check, etc.
+transpileExpr :: Expr_ -> State Context [Instruction]
```
Comment (wingedtrashcan) [inline src/Nightfall/Targets/Miden.hs:144]: So the denotation of the given `Expr_` is equal to the top value on the stack after the instructions returned by `transpileExpr` are evaluated in the returned context, regardless of what the initial context was. Is that correct? If so, we should make a property test of that in some future.

**Review by wingedtrashcan — COMMENTED** [review]
Another chunk. Sorry about the amount of random questions and comments I've been dumping on you, I promise I'll shut up at some point :)

File: src/Nightfall/Lang/Internal/Types.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (wingedtrashcan) [reply src/Nightfall/Lang/Internal/Types.hs:33]: > If you have a smart constructor for each specific builtin, you can "recover" precise types I guess?

Yes, but only at construction time. I.e. when processing the AST you need to make sure to preserve the length of the list of arguments and the type checker won't help you (unless we use a GADT there). Or to allow for under-application (i.e. partial application) and over-application (i.e. applying a function returned by a function) and handle those everywhere. Miden obviously doesn't need the latter, but depending on how high-level we want to make Nightfall, we may find those useful.

I think we have enough discussion here to have a dedicated GitHub issue, so I'll create one.

File: src/Nightfall/MASM/Types.hs (hunk 13 above)
Comment (GoNominal) [reply src/Nightfall/MASM/Types.hs:75]: > Not sure if I'm being crazy, but I thought those were field operations? I.e. defined over Felt, which is Word64 rather than Word32?

No you're not crazy, they _are_ field operations indeed :-)
I guess I just went a little too fast when made the modification to add `Mul` and `Div`. I went ahead an mimicked the `Add` and `Sub` that were there.
Come to think of it, we might have made some confusion there, because sometimes the argument is an index in memory (`.i`) and sometimes it's a field element (`.n`).

So that's a typo here, and should definitely be a `Word64` because `Add` and co takes an optional field element.

Good catch!

Comment (GoNominal) [reply src/Nightfall/MASM/Types.hs:75]: As for the others, `Push` indeed takes a `Word64` because it pushes a field element on the stakc, but `Swap` stays with a `Word32`as the argument is an index within the stack, to decide which element to swap with.

_Actually_ `swap` only accepts an argument between `1` and `15` as only the first `16` elements of the stack are reachable and it doesn't make sense to swap the top of the stack with itself.
So we could even make a new type for "Stack Index" if we want.

File: src/Nightfall/MASM/Types.hs (hunk 12 above)
Comment (GoNominal) [reply src/Nightfall/MASM/Types.hs:79]: Well, it actually comes down to history: all the Miden part of the code comes from an earlier internal project in which @allezlom13 wrote a small DSL for Miden. We re-used this existing code in `Nightfall` because it made sense.
But the two project's goals were different. In the former, I guess we were trying to be closer to Miden's inner working ; this is why the DLS allowed to generate a broader subset of Miden's assembly.

When I started `Nightfall` and added Miden as a backend, the goal was to produce efficient Miden goal from the DSL. I did not need (nor did I want to) make a 100% fully-compatible Miden's DSL.

So my train of thoughts at the time were: "if the program in Nightfall adds to 32-bit integers, I should generate the Miden code to add them". So in this case I don't _need_ to support both ways of adding 32-bit integers. Besides, the [doc](https://wiki.polygon.technology/docs/miden/user_docs/assembly/u32_operations#arithmetic-operations) says that the variant with the `.b` argument takes `1-2` more cycles.
So I settled for always using the "naked" version, hence why I didn't need the `Maybe Int32` variant.

I guess it comes down to the fact that I'm using a _partial_ Miden DSL inside Nightfall, just supporting what I need, and I wasn't trying to provide a full Miden DSL.

I would even argue that we could remove all the variants with a `Maybe` Argument, because I don't think we will ever benefit from using them?

But I might be wrong, this is still early.

To be discussed!

File: app/Main.hs (hunk 8 above)
[the remark this answers is not in the document]
Comment (GoNominal) [reply app/Main.hs:35]: > I'd do it this way too probably. No matter how many times I used optparse-applicative, I still have to relearn it completely from scratch every single time.


Yeah, there's definitely that :D

But when we're ready to have the executable "as we want it" (take program as input, generate the backend, etc.) it makes sense to use `optparse-applicative`. 100%.

File: src/Examples/Fib.hs
[the remark this answers is not in the document]
Comment (GoNominal) [reply src/Examples/Fib.hs:1]: Yes indeed ^^
Sorry about that.

File: src/Examples/Simple.hs (hunk 10 above)
[the remark this answers is not in the document]
Comment (GoNominal) [reply src/Examples/Simple.hs:66]: Np.

To be fair, I think those files / modules in `Examples/` would benefit from being `.lhs` files with the tutorial written directly in.
Like "here is how to use the most basic constructs: adding two ints" and then show the `trivial1Stmts` or whatever it is called.

Then going on with "a multiplication would look like this:" (show code), and then "we can get rid of those pesky `lit` constructs and those `add`," (and then shoe the uncluttered version), etc.

That would make more sense to treat those modules as sort of "linear" from top to bottom. Which is how I built them anyway ^^

File: src/Nightfall/Lang/Internal/Types.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (GoNominal) [reply src/Nightfall/Lang/Internal/Types.hs:33]: Ok perfect. I'm closing this in favor of the issue then.

File: src/Nightfall/Lang/Internal/Types.hs (hunk 1 above)
Comment (GoNominal) [reply src/Nightfall/Lang/Internal/Types.hs:68]: Leaving it up to you as how you treat it. Obviously not a priority, but if it's fast enough you can totally implement the change.

Closing this as I believe you know how to proceed with this one :)

File: src/Nightfall/Lang/Types.hs (hunk 2 above)
Comment (GoNominal) [reply src/Nightfall/Lang/Types.hs:109]: An yes sorry this makes sense that `add True 67` would not typecheck and doesn't have anything to do with the `Num` constraint.
Sometimes lack of caffeine ....

File: src/Nightfall/Lang/Types.hs (hunk 3 above)
Comment (GoNominal) [reply src/Nightfall/Lang/Types.hs:186]: Ah yes ^^

File: src/Nightfall/Targets/Miden.hs [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+    warning <- transpileStatement . coerce $ comment "This program was generated by Nightfall (https://github.com/qredo/nightfall), avoid editing by hand.\n"
+    modify $ \ctx -> ctx { progName = pName zkProg }
+    midenInstr <- concat <$> mapM (transpileStatement . coerce) (pStatements zkProg)
+    return $ Module { moduleImports = [] -- No import for now (TODO)
+                    , moduleProcs = Map.empty -- No procs either (TODO)
+                    , moduleProg = Program (head warning : midenInstr)
+                    }
+
+transpileStatement :: Statement_ -> State Context [Instruction]
+transpileStatement (NFTypes.Comment str) = return . singleton . MASM.Comment . Text.pack $ str
```
[the remark this answers is not in the document]
Comment (GoNominal) [reply src/Nightfall/Targets/Miden.hs:68]: Totally. Would prevent the `Text.pack` yeah.

File: src/Nightfall/Targets/Miden.hs (hunk 4 above)
Comment (GoNominal) [reply src/Nightfall/Targets/Miden.hs:70]: Ah that's my lack of Haskell experience talking then.
It it exists, then let's use it.
It's curious that HSL didn't suggest it.

File: src/Nightfall/Targets/Miden.hs (hunk 15 above)
Comment (GoNominal) [reply src/Nightfall/Targets/Miden.hs:73]: Sorry I'm not sure I caught this one?

File: src/Nightfall/Targets/Miden.hs (hunk 14 above)
Comment (GoNominal) [reply src/Nightfall/Targets/Miden.hs:74]: This one might be better off in its own issue? I think it's an important point.

The initial goal (which might move since we have already talked about it earlier) was to provide a "write once, use many times" approach with this eDSL.

_i.e._ users writing their ZK algorithm should not care about the underlying backend. The purely stack-based approach is specific to Miden and others might not be.
IMO it makes sense to provide a high-level, typed `If` construct to the user for branching.

That's a discussion of course and we might ditch this, but I think it makes sense. The goal was to open ZK world to non-zk specialists.

As for having this optimized, I can see several solutions :

- giving the ability to write inline native code (like wriring inline assembly in C code for instance). That would obviously break the generality, but _if_ that's what the users wants to do, he's obviously aware of the backend.
That's a possibility, not sure this is (easily) doable. I'm not a fan

- we could make nano-passes and detect this? For instance, we could either detect than an `IfElse` block uses the previously-returned value, and short-cut the generated code ?

File: src/Nightfall/Targets/Miden.hs (hunk 5 above)
Comment (GoNominal) [reply src/Nightfall/Targets/Miden.hs:88]: It certainly doesn't come from Miden as it doesn't even have the concept of variables.
I have introduced the concept of variables in `Nightfall` and decided to map them to Miden's global memory location (this in itself could be an issue to talk about, because I have some ideas / alternatives to that that would provide some performance gains, but require a bit more work on our side).

The fact that I didn't allow shadowing is two-folds:

- To write this eDSL I took inspiration from an eDSL I wrote earlier to generate C code, so I might have been a little fast in my thinking
- Again due to time constraints (I want to have a PoC working "soon") I did not want to have to think about scopes, etc. So I decided not to allow them: faster for me to implement, less error-prone. And no real downside save for the user having to use another name explicitly.
I don't think it's much of an issue for loop if you just define the variable outside of it. It's possible to rewrite a variable, so using them in loop is not a problem. Only redefining them is not allowed.

Does that make sense?

File: src/Nightfall/Targets/Miden.hs (hunk 6 above)
Comment (GoNominal) [reply src/Nightfall/Targets/Miden.hs:94]: Not as of yet no.
For now, memory names are really directly mapped to a global memory location. I haven't defined any concept of scope.

File: src/Nightfall/Targets/Miden.hs (hunk 16 above)
Comment (GoNominal) [reply src/Nightfall/Targets/Miden.hs:130]: Yes, I tend to go for `String` first and then refactor. This is a bad habit.
`Text` isn't that much harder, especially with `OverloadedStrings`.

Fee free to change that :)

File: src/Nightfall/Targets/Miden.hs (hunk 15 above)
Comment (wingedtrashcan) [reply src/Nightfall/Targets/Miden.hs:73]: I think Miden allows an `if` statement without an `else` clause? Oh! I'm just being stupid, we represent those as `ifElse` with an empty list of statements for the `else` clause and so we don't need any `Maybe` in there. Sorry for the noise.

File: src/Nightfall/Targets/Miden.hs (hunk 16 above)
Comment (wingedtrashcan) [reply src/Nightfall/Targets/Miden.hs:130]: > Yes, I tend to go for `String` first and then refactor. This is a bad habit.

Yeah, I also have that habit. `import Data.Text` introduces lots of name clashes and so one needs to use `qualified`, but then `Text.Text` is stupid and so that's another `import Data.Text (Text)` and then all functions over `Text` need to be used qualified... Bad UX really can get in the way.

> Fee free to change that :)

Will do.

File: src/Nightfall/Lang/Internal/Types.hs (hunk 11 above)
[the remark this answers is not in the document]
Comment (wingedtrashcan) [reply src/Nightfall/Lang/Internal/Types.hs:33]: > I'm closing this in favor of the issue then.

[Done](https://github.com/qredo/nightfall/issues/17#issue-1821266909).

File: src/Nightfall/Targets/Miden.hs (hunk 5 above)
Comment (wingedtrashcan) [reply src/Nightfall/Targets/Miden.hs:88]: > It's possible to rewrite a variable

Right, purity is way to wired into my brain at this point lol.

> Does that make sense?

Yes, makes perfect sense, thanks for the explanation!

File: src/Nightfall/MASM/Types.hs (hunk 12 above)
Comment (wingedtrashcan) [reply src/Nightfall/MASM/Types.hs:79]: > Besides, the [doc](https://wiki.polygon.technology/docs/miden/user_docs/assembly/u32_operations#arithmetic-operations) says that the variant with the .b argument takes 1-2 more cycles.

I guess it's because `push` is itself 1-2 cycles, so the `.b` version probably just performs a `push`.

> I would even argue that we could remove all the variants with a `Maybe` Argument, because I don't think we will ever benefit from using them?

Sounds like the `.b` version of, say, `add` might be more optimal in case of field operations? It can be 1 (at most 2) cycle vs non-`.b` version being still 1 cycle but requiring an additional `push`, which is also 1-2 cycles. 

So I agree it doesn't seem like we need them for 32-bits arithmetic, but I'd keep existing ones in, since they're already there and may potentially be beneficial in future.

File: src/Nightfall/MASM/Types.hs (hunk 13 above)
Comment (wingedtrashcan) [reply src/Nightfall/MASM/Types.hs:75]: > Actually swap only accepts an argument between 1 and 15 as only the first 16 elements of the stack are reachable and it doesn't make sense to swap the top of the stack with itself.
> So we could even make a new type for "Stack Index" if we want.

Right, great point.

OK, if you don't mind, then my next PR is going to be a cleanup of various word-like types. And I'll make `Felt` a `newtype`, again if you don't mind.
